src/Stomp/Network/Observer/Heartbeat/Emitter.php

```diff
@@ -0,0 +1,260 @@
+     * @var float
+     */
+    private $interval = null;
+
+    /**
+     * Whenever the emitter is configured to send beats.
+     *
+     * @var bool
+     */
+    private $enabled = false;
+
+    /**
+     * Defines the percentage amount of the calculated interval that will be used without emitting a beat.
+     *
+     * @var float
+     */
+    private $intervalUsage = 0.65;
+
+    /**
+     * Emitter constructor.
+     * @param Connection $connection
+     */
+    public function __construct(Connection $connection)
+    {
+        $this->connection = $connection;
+    }
+
+
+    /**
+     * Indicates that during a read call no frame was received, but an EOL line.
+     *
+     * @return void
+     */
+    public function emptyLineReceived()
+    {
+        $this->onPassiveEvent();
+    }
+
+    /**
+     * An passive event was fired, all that triggers a connection activity that is not leading to outgoing traffic.
+     *
+     * @return void
+     */
+    private function onPassiveEvent()
+    {
+        if ($this->isDelayed()) {
+            $this->sendBeat();
+        }
+    }
+
+    /**
+     * Check if the emitter is in a state that indicates a delay.
+     *
+     * @return bool
+     */
+    public function isDelayed()
+    {
+        if ($this->enabled) {
+            $now = microtime(true);
+            return ($now - $this->lastbeat > $this->interval);
+        }
+        return false;
+    }
+
+    /**
+     * Send a beat to the server.
+     *
+     * @return void
+     */
+    private function sendBeat()
+    {
+        $this->connection->sendAlive();
+        $this->notifyBeat();
+    }
+
+    /**
+     * Outgoing activity event.
+     *
+     * @return void
+     */
+    private function notifyBeat()
+    {
+        $this->lastbeat = microtime(true);
+    }
+
+    /**
+     * Indicates that during a read call no frame was received, but an EOL line.
+     *
+     * @return void
+     */
+    public function emptyBuffer()
+    {
+        $this->onPassiveEvent();
+    }
+
+    /**
+     * Indicates that a frame has been received.
+     *
+     * @param Frame $frame that has been received
+     * @return void
+     */
+    public function receivedFrame(Frame $frame)
+    {
+        if ($frame->getCommand() === 'CONNECTED') {
+            $beats = $this->getHeartbeats($frame);
+            $this->intervalServer = $beats[1];
+            if ($this->intervalServer && ($this->intervalClient || $this->intervalClient === null)) {
+                $this->interval = $this->intervalUsage * (max($this->intervalServer,
+                            $this->intervalClient) / 1000); // milli to micro
+                $this->enabled = true;
+            }
+        } else {
+            $this->onPassiveEvent();
+        }
+    }
+
+    /**
+     * Returns the heart beat header.
+     *
+     * @param Frame $frame
+     * @return array
+     */
+    private function getHeartbeats(Frame $frame)
+    {
+        $beats = $frame['heart-beat'];
```

Comment:
The `ArrayAccess` implementation of `Frame` will return `null` if the header is not available. That simplified some work with the headers.